Mpeg2decoder: fixes to QuantMatrices and NalReader redefine

chivakker opened the pull request:
Mpeg2decoder adds support to updating Quantization Matrices.  Also NalReader redesign helped to remove arithmetic on the StarCodePrefixSize.

chivakker committed:
mpeg2decoder: properly parse and update Quant Matrix Extension

Quant Matrix Extension comes within the stream and can update one
or more matrices.  The Sequence Header can also update matrices to
default values or values from stream and it has to start from a
clean reference

Signed-off-by: Daniel Charles <daniel.charles@intel.com>

chivakker committed:
mpeg2dec: remove arithmetic computations with redefined value

kStartCodePrefixSize was redefined to 0 per NalReader redesign,
so there's no need to do arithmetic computations with it.

Signed-off-by: Daniel Charles <daniel.charles@intel.com>

chivakker committed:
mpeg2: after decoding a picture there's no sliceparsing

When a sequence end code comes and the current picture is
decoded parsing slices is not happening anymore

Signed-off-by: Daniel Charles <daniel.charles@intel.com>

xuguangxin commented on decoder/vaapidecoder_mpeg2.cpp:
```diff
@@ -628,31 +703,74 @@ YamiStatus VaapiDecoderMPEG2::assignPicture()
 }
 
+void
+VaapiDecoderMPEG2::updateIQMatrix(const QuantMatrices* refIQMatrix, bool reset)
```
may need format code using clang-format

xuguangxin commented on codecparsers/mpeg2_parser.cpp:
```diff
@@ -356,12 +355,96 @@ namespace MPEG2 {
+            return false;
+        }
+
+        bitReaderReadFlag(&(quantMatrices->load_intra_quantiser_matrix));
```
Could we use one function like
loadMatrix(bool& loaded, uint8_t matrix[64])
to remove 4 duplicate code below?

xuguangxin commented on codecparsers/mpeg2_parser.cpp:
```diff
@@ -356,12 +355,96 @@ namespace MPEG2 {
+            // read 8 bits *64 from the stream
+            uint32_t value;
+            for (uint8_t i(0); i < 64; ++i) {
+                bitReaderReadBits(8, &value);
```
better check we have enough data to read or not.

chivakker replied:
I'll put a check for it

xuguangxin commented on codecparsers/mpeg2_parser.cpp:
```diff
@@ -568,36 +646,40 @@ namespace MPEG2 {
         bitReaderReadFlag(&(m_sequenceHdr.constrained_params_flag));
 
-        bitReaderReadFlag(&(m_sequenceHdr.load_intra_quantiser_matrix));
+        bitReaderReadFlag(&(quantMatrices->load_intra_quantiser_matrix));
```
better use following function to remove duplication
`loadMatrix(bool& loaded, uint8_t matrix[64], uint8_t defaultMatrix[64])`
and this function can call 
`loadMatrix(bool& loaded, uint8_t matrix[64])`
to remove duplicate code further

chivakker replied:
I'll refactor this and resubmit.

xuguangxin commented on decoder/vaapidecoder_mpeg2.cpp:
```diff
@@ -400,36 +438,73 @@ YamiStatus VaapiDecoderMPEG2::processDecodeBuffer()
+            if (status != YAMI_SUCCESS) {
+                return status;
+            }
+            status = m_DPB.outputPreviousPictures(m_currentPicture, true);
```
from 

> Figure 6-14. High level bitstream organisation.

picture data will switch to 4 dest.
1. \- sequnce extension
2. \- group picture header
3. \- sequnce end
4. \- picture coding extension

do we need output previous picutre for 1, 2?

chivakker replied:
I am not sure I follow here, but yes picture 1, 2 are output as the sequence ended.

xuguangxin replied:
Ok, but I still have other related question. 1, 2(with close gop) and 3  should flush the reference frame in dpb, right? It will avoid frames in next gop reference previous gop. But I did not see reference frame clean logical in the code.

xuguangxin replied:
You may miss this comment. Did we flush dpb for 1, 2, 3 or not?From my memory we should not reference previous picuture any more if we see 1,2, 3. Correct me if i am wrong.
thanks

chivakker commented:
Resubmitted with the solicited refactor done.

xuguangxin commented on codecparsers/mpeg2_parser.cpp:
```diff
@@ -568,36 +633,22 @@ namespace MPEG2 {
-                m_sequenceHdr.non_intra_quantiser_matrix[i] = value;
-            }
-        } else {
+        readQuantMatrix(quantMatrices->load_non_intra_quantiser_matrix,
```
do you think it's ok to add other readQuantMatrix function with 3 params.
readQuantMatrix(bool& loadMatrix, uint8_t matrix[], uint8_t defaultMatirx[]).
 it will reduce your logical code.

chivakker replied:
There's a trade off, the only place where this logic is needed is here and the matrices can be identified properly here (default quant matrix and all 16 non-intra matrix), that way I don't need to pass it by value to the next function.

xuguangxin replied:
Is it ok to define kDefaultNonIntraBlockMatrix just as you define kDefaultIntraBlockMatrix, Although it all 16. But you can share some function as readQuantMatrix(bool& loadMatrix, uint8_t matrix[], uint8_t defaultMatirx[]). 
The basic idea is If you look at L636 to L643.They are same logical as L646 to  L651. right?

xuguangxin commented on codecparsers/mpeg2_parser.cpp:
```diff
@@ -356,12 +355,83 @@ namespace MPEG2 {
+    void Parser::readQuantMatrix(bool& loadMatrix, uint8_t matrix[])
+    {
+
+        bitReaderReadFlag(&loadMatrix);
```
Better check return value, what if not enough data from bitstream? Bitstream valid check is important and more important to mpeg2, Because mpeg2 usually from  cable network, it may introduce error.

chivakker replied:
So far the nal units will be checked for the smallest size they can have, a broken nal unit will result in artifacts or complete  shutdown of the component depending on what is missing from the stream.  Some of them will be non-fatal and decoding can continue. If you agree, a more strict size enforcement can be made in the future with a good categorization on the nature of the missing information.  It will be bad to always shutdown if information is missing on the stream.

xuguangxin replied:
ok, maybe we can report no fatal error in future

xuguangxin commented on codecparsers/mpeg2_parser.cpp:
```diff
@@ -356,12 +355,83 @@ namespace MPEG2 {
+            return false;
+        }
+
+        readQuantMatrix(quantMatrices->load_intra_quantiser_matrix,
```
better check return value.

chivakker committed:
mpeg2dec_unittest: fix use cases for new nalReader

nalReader is operating without the start code prefix so
the unittest has to be adapted to that.

Fixes #479

Signed-off-by: Daniel Charles <daniel.charles@intel.com>

jayesht commented:
I verified this pull request on Skylake:
Linux 4.4.0-24-generic #43-Ubuntu SMP Wed Jun 8 19:27:37 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux with the following patches:
01e4a09 mpeg2dec_unittest: fix use cases for new nalReader
cd83e25 mpeg2: after decoding a picture there's no sliceparsing
bd09c63 mpeg2dec: remove arithmetic computations with redefined value
8d6b8cd mpeg2decoder: properly parse and update Quant Matrix Extension

mpeg2 unittests are now passing with 37/39, 2 failing tests are:
twilight_zone_mei_MEI2.stream17.mpeg2    fail
twilight_zone_mei_MEI.stream17.long.mpeg2    fail

I would redo it when there is a new revision submitted for the patches.

chivakker committed:
mpeg2decoder: flush dpb with a new sequence header

sequence header should start with an empty dpb

Signed-off-by: Daniel Charles <daniel.charles@intel.com>

jayesht commented:
I verified this pull request on Skylake:
Linux 4.4.0-24-generic #43-Ubuntu SMP Wed Jun 8 19:27:37 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux with the following patches:

b31a362 mpeg2decoder: flush dpb with a new sequence header

472087a mpeg2dec_unittest: fix use cases for new nalReader
2ffe38e mpeg2: after decoding a picture there's no sliceparsing
57d8738 mpeg2dec: remove arithmetic computations with redefined value
d7a8535 mpeg2decoder: properly parse and update Quant Matrix Extension

mpeg2 unittests are now passing with 37/39, 2 failing tests are:
twilight_zone_mei_MEI2.stream17.mpeg2 fail
twilight_zone_mei_MEI.stream17.long.mpeg2 fail

xuguangxin commented:
ok +1, hope we can address sequnce and gop things later.
